## **Analysis of Task 2 Results:**

**Precision and recall of the SNP caller**

Precision is given by TP/(TP+FP)

Recall is given by TP/(TP+FN)

[EcoliK12-MG1655:]{.underline}

**For Indels** 100% of True Positives were identified by the first part (bcftools) of the task so a **precision** of 1 and a **recall** of 1. In the pipeline bcftools identified all INDELs with no False Positives or False Negatives. Snippy missed one INDEL therefore has a **precision** of 1 and a **recall** of 0.95.

**For SNPs** 295 of the 300 SNPs were found by the first part (bcftools) of the task. It found 4 False Positives hence a **precision** of 0.99. It found 5 False Negatives so a **recall** of 0.98. In the pipeline bcftools found 295 of the 300 SNPs with no False Positive and 5 False Negatives. Therefore a **precision** of 1 and a **recall** of 0.98. Snippy found 291 of the 300 SNPs with no False Positives and 9 False Negatives. Therefore a **precision** of 1 and a **recall** of 0.955.

[NC-037282.1:]{.underline}

**For Indels** 100% of True Positives were identified by the first part (bcftools) of the task so a **precision** of 1 and a **recall** of 1. In the pipeline bcftools identifed all indels with no FP or FN. Snippy missed one indel therefore has a **precision** of 1 and a **recall** of 0.95.

**For SNPs** 298 of the 300 SNPs were found by the first part (bcftools) of the task. It found 8 False Positives hence a **precision** of 0.97. It found 2 False Negatives so a **recall** of 0.99. In the pipeline bcftools found 298 of the 300 SNPs with no False Positive and 2 False Negatives. Therefore a **precision** of 1 and a **recall** of 0.99. Snippy found 294 of the 300 SNPs with no False Positives and 9 False Negatives. Therefore a **precision** of 1 and a **recall** of 0.97.

On the simulated data both perform well. Just by looking at the NC-03782.1.fasta it has noticeably more repeats at the beginning of the genome than the EcoliK12-MG1655.fasta which may have caused alignment problems with more mutations in this part of the genome. Recall is slightly poorer for Snippy than for bcftools. The combined VCF does not perform significantly better.

[Real Data:]{.underline}

The Reference Genome was EcoliK12-MG1655.fasta. The fastq.gz files were SRR25083113_1.fastq.gz and SRR25083113_2.fastq.gz

| Indels             | Snippy | bcftools |
|--------------------|--------|----------|
| Total              | 8700   | 547      |
| Call not confident | 0      | 69       |
| Unique to caller   | 8359   | 203      |

| SNPs               | Snippy | bcftools |
|--------------------|--------|----------|
| Total              | 47207  | 69539    |
| Call not confident | 0      | 1306     |
| Unique to caller   | 197    | 22332    |

There is a clear difference in the performance of bcftools and Snippy. Snippy has called more INDELs than bcftools however, bcftools has recognized more SNPs than Snippy. Snippy was more confident in aligning reads.

Included are the Depth coverage for reads generated by bcftools and Snippy. This shows that bcftools bam file has aligned a significantly larger number of reads at the mid 1kb point of the genome than the Snippy bam file.

Snippy is designed for bacterial genomes. Whereas the BCFtools is a more general purpose variant caller. Bacterial genomes have shorter INDELS in general so Snippy may have identified regions that were ignored or called as SNPs by BCFtools. INDELs could also be interpreted as a number of SNPs close together by the more generalized tool. An example of this is POS 1347 where Snippy has called an INDEL REF: AGCG ALT:TGCA whereas BCFtools has called a SNP POS 1347 REF:A ALT:T and POS 1350 REF:G ALT:A which is essentially the same mutation.

I calculated the average read depth from the VCF files. Snippy was 82.7364 and BCFtools was 94.9138

**Scoring the variant callers:\
SNPs**

concordance = (Total SNPS - Not confident reads)/(Total SNPs)

For Snippy this is 1

For BCFtools this is 0.981

Normalize read depths: Snippy = (82.7364/94.9138) = 0.872 and BCFtools = (94.9138/94.9138) = 1

Score = (Normalized depth) \* (Concordance)

Snippy_SNPs_score = 0.872 BCFtools_SNPs_score = 0.981

**INDELS**

concordance = (Total INDELs - Not confident reads)/(Total INDELs)

For Snippy this is 1

For BCFtools this is 0.873

Snippy_INDEL_score = 0.872 BCFtools_INDEL_score = 0.873

This suggests that BCFtools may have a similar quality of INDEL alignment but superior SNP alignment than Snippy. However, the spike in read depth in the BCFtools VCF may skew these scores. With the same number of reads in the fastq files a lower depth might actually be accurate unless there is a significant area of repeats in the genome.

[**Future improvements to code:**]{.underline}

When creating the dictionaries I would have converted them to VCF files rather than JSON files so that I could run bcftools isec on the data and get full statistics rather than convert to CSV and analyse in Excel. I realized this too late in the process.

I would also have created a variable that would have identified if the input was single end or paired end reads and put this into the pipeline where I run Snippy.

Before merging I would have normalized my VCF files to try and get the INDELS a more comparable format. With the simulated data I did the analysis by eye as there were only 20 INDELs but this would not be possible for larger data sets.

**Associated Files:**

Included are the Python files with the code. Excel files with all the results from the run of the python files. The depth plot files for BCFtools and Snippy. I created a Conda environment called varcalltask2. There are some additional files in the directory that were used for experiments.

**Directory:\
**(varcalltask2) jovyan:/shared/team/people/luke/Task2/Folder_containing_files_for_Task2
